hip-config.cmake.in

```diff
@@ -120,13 +120,19 @@ else()
 endif()
 
 if(HIP_COMPILER STREQUAL "clang")
-  if(WIN32)
-    set(HIP_CLANG_ROOT "${HIP_PATH}")
-  else()
-    set(HIP_CLANG_ROOT "${ROCM_PATH}/llvm")
+  if(NOT HIP_CLANG_ROOT)
+    if(WIN32)
+      set(HIP_CLANG_ROOT "${HIP_PATH}")
+    else()
+      set(HIP_CLANG_ROOT "${ROCM_PATH}/llvm")
+    endif()
   endif()
   if(NOT HIP_CXX_COMPILER)
-    set(HIP_CXX_COMPILER ${CMAKE_CXX_COMPILER})
+    if(CMAKE_HIP_COMPILER_WITH_PATH)
```

Comment:
What is `CMAKE_HIP_COMPILER_WITH_PATH`? That looks like the compiler used for HIP language in cmake, which should not be used here.  `HIP_CXX_COMPILER` should not be using the compiler for the HIP language. hip-config.cmake is for the CXX language(and flags are restricted to only apply to the CXX language) and therefore should use the `CMAKE_CXX_COMPILER`. If we implicitly use `CMAKE_HIP_COMPILER` it can cause problems when user links with `hip::device` and the user is using a different `CMAKE_CXX_COMPILER`.

This variable `HIP_CXX_COMPILER` is only meant as a workaround in MIOpen in order to serialize the flags from `hip::device` but not link with this target.

Reply:
Even when I use HIP as a language CMake, I need to invoke find_package(hip) otherwise, libraries like hipblas will not be happy. Since HIP_CLANG_ROOT is derived from HIP_CXX_COMPILER, Need to teach HIP_CXX_COMPILER  using CMAKE_HIP_COMPILER_WITH_PATH which will be set by hip-lang-config.cmake which is invoked before hip-config.cmake. If you don't use hip-lang.config, it is no-op here.

Reply:
Intended to get the following works nicely.
```
  enable_language(HIP)
  list(PREPEND CMAKE_PREFIX_PATH ${CMAKE_HIP_COMPILER_ROCM_ROOT})
  find_package(hip CONFIG REQUIRED)
  find_package(hipblas CONFIG REQUIRED)
```
`enable_language(HIP)` invokes `hip-lang-config.cmake`

Reply:
I think I got what you mean. Unfortunately, is not easy to set a clear boarder between HIP and CXX and that is why I'm having issue with source code being both HIP and need CXX host library. Need a bit more thinking.

Reply:
When I use HIP as language, I'd like to compile with CMAKE_HIP_COMPILER and link with CMAKE_CXX_COMPILER and there is only one hip runtime being linked in preferably the one favored by CMAKE_HIP_COMPILER. I also call host hip libraries and that is why there is entanglement.

Reply:
> Right now Clang_rt pollution comes into my linking via hip::host.

I see. That looks wrong. The comment mentions its needed for __fp16 or _Float16 types, but those are used for device compilation and not host-side API. If its needed to use our host APIs, then that means users must use hip clang for both CMAKE_CXX_COMPILER and CMAKE_HIP_COMPILER. Hopefully thats not the case. Maybe @yxsamliu knows more.

If hip::host does not need a device compiler then we can remove the clang_rt flags from hip::host.

> If they are need by MIOpen only, should it be moved to MIOpen CMake instead of here?

The HIP_CXX_COMPILER flags is to get the compiler flags for hip::device(not for hip::host). And MIOpen is only serializing the hip::device flags not the hip::host flags.